File: dispatcher/dispatcher.php
<?php
/************************
dispatcher.php
This process continuously checks whether there exist unjudged submissions.
If there's unjudged submissions, start 'submit_curl.php' process to send it to a available judge VM.
**************************/


require_once("dispatcherFunction.php");


writeLog("================================");
writeLog("================================");

$machine = getMachine();
if($machine!=null)
	initMachine($machine);
else
	writeLog("get machine file error");
print_r($machine);

$i=0;
echo "submit_test start<br>";
writeLog("dispatcher start");
while(1)
{
	/*sleep(10);*/
	$con = get_database_object();
    $sidQuery = "SELECT * FROM problem_submission WHERE status = 'WAIT' ORDER BY id ASC LIMIT 100";
    $sidRs = mysql_query($sidQuery) or die(mysql_error());
	
	echo "sidRsNum".mysql_num_rows($sidRs)."\n";
	while( $sidRow = mysql_fetch_array($sidRs) ) {
		$sid = $sidRow['id'];	
		$pid = $sidRow['problem_id'];
		$problemQuery = "SELECT * FROM problem_problem where id = '$pid'";
		$problemRs = mysql_query($problemQuery);
		$problemRow = mysql_fetch_array($problemRs);
		$judge_source = $problemRow['judge_source'];
		$outside_web = $problemRow['judge_type']; 
		$codeLanType = $sidRow['language'];
		
		if($judge_source == "LOCAL")
		{	
			$findMachine = 0;
			echo "before findmachine\n";
			writeLog("before find machine");
			$destMachine = null;
			while($destMachine==null)
			{
				$destMachine = getIdleMachine($machine);
				sleep(1);
			}

			if($destMachine != null)
			{
				echo "destMachineName=".$destMachine['machineName']."   destMachineIP=".$destMachine['machineIP']."\n";
				writeLog("submission ".$sid." use ".$destMachine['machineName'].":".$destMachine['machineIP']."to judge");
			}
			else
			{
				echo "destMachine=null!!\n";
				writeLog("find machine error");
			}
		
			$con = get_database_object();
			$updQuery = "UPDATE problem_submission SET status = 'JUDGING' WHERE id = ".$sid;
			echo $updQuery."   ".$sid."\n";
			mysql_query($updQuery) or die(mysql_error());
			$judgeIP = $destMachine['machineIP'];
			$judgePage = "interface.php";
			$judgeURL = $judgeIP."/".$judgePage;
			echo "judgeURL=".$judgeURL."\n";	
		
			echo "send!\n";

			writeLog("send info to judge ".$destMachine['machineName'].":".$destMachine['machineIP']);
			$arg = $sid." ".$pid." ".$codeLanType." ".$destMachine['machineName']." ".$judgeURL;
			$handle = popen("/usr/bin/php submit_curl.php ".$arg." & >> output.html", "w");
			pclose($handle);
			echo "after popen\n";
		}
		else
		{
			$con = get_database_object();
            $updQuery = "UPDATE problem_submission SET status = 'JUDGING' WHERE id = ".$sid;
            echo $updQuery."   ".$sid."\n";
            mysql_query($updQuery) or die(mysql_error());
			
			writeLog("send info to other judge");
			$arg = $codeLanType." ".$pid." ".$outside_web." ".$sid;
			if(file_exists("/var/nthuoj/outsideConnection/sendToOtherJudge.sh"))
			{
				$handle = popen("/var/nthuoj/outsideConnection/sendToOtherJudge.sh ".$arg." & >> output.html", "w");
				pclose($handle);
			}
			else
				writeLog("send info to other judge error(no judge)");

		}
	}

	sleep(10);
	$i++;
}
echo "submit_test finished\n";
writeLog("dispatcher finished");
?>

File: ojbackend/dispatcher.py
"""
    dispatcher.py
    This process continuously checks whether there exist unjudged submissions.
"""

from dispatcherFunc import *
import time
import logging
import subprocess, sys
logging.basicConfig(filename = 'dispatcher.log', 
                    level = logging.INFO, 
                    format = '%(asctime)s ::%(message)s',
                    datafmt = '%m/%d/%Y %I:%M:%S %p')
logging.info('==========Dispatcher Started==========')

machineInfo = getMachine()
if machineInfo != None:
    initMachine(machineInfo)
    logging.info('getMachine Success')
else:
    logging.info('getMachine Error')
    logging.info('Please check the file machineInfo.config to check the settings')
    logging.info('==========Dispatcher Finished==========')
    exit(0)

dbIP, dbUser, dbPasswd, dbName = getdbInfo()
if dbIP == None or dbUser == None or dbPasswd == None or dbName == None:
    logging.info('getdbInfo Error')
    logging.info('Please check the settings in ojdatabase.config')
    logging.info('==========Dispatcher Finished==========')
    exit(0)
logging.info('connect to database')

while True:
   
    DB = connectDB(dbIP, dbUser, dbPasswd, dbName)
    if DB == None:
        logging.info('connect database error')
        logging.info('Please check the settings in ojdatabase.config')
        logging.info('==========Dispatcher Finished==========')
        exit(0)
    submissionStat = " \
    SELECT * FROM problem_submission \
    WHERE status = \'WAIT\' \
    ORDER BY id ASC \
    LIMIT 100; \
    "
    cur = DB.cursor()
    cur.execute("USE nthuoj;")
    cur.execute(submissionStat)
    sidQuery = cur.fetchone()
    
    while sidQuery != None:
        sid = sidQuery[0]
        pid = sidQuery[1]
        logging.info('sidQuery Success!')
        logging.info('sid = %d, pid = %d' % (sid, pid))
        problemID = " \
        SELECT * FROM problem_problem where id = \'%d\'; \
        "
        cur.execute(problemID % pid)
        pidQuery = cur.fetchone()
        judgeSource = pidQuery[11]
        judgeType = pidQuery[12]
        judgeLanguage = pidQuery[13]
        logging.info('pidQuery Success!')
        logging.info('judgeSource = %s, judgeType = %s, judgeLanguage = %s' 
                    % (judgeSource, judgeType, judgeLanguage))

        if judgeSource == "LOCAL":
            idleMachine = None
            logging.info('get idleMachine')
            while idleMachine == None:
                time.sleep(1)
                idleMachine = getIdleMachine()       
            logging.info('idleMachine = %s' % idleMachine) 
            updatesidStat = " \
            UPDATE problem_submission \
            SET status = \'JUDGING\' \
            WHERE id = \'%d\'; \
            "
            cur.execute(updatesidStat % sid)
            DB.commit()
            judgeIP = machineInfo[idleMachine]
            judgeURL = judgeIP + "/interface.py"
            logging.info('get judgeURL = %s' % judgeURL)            
        else :
            updatesidStat = " \
            UPDATE problem_submission \
            SET status = \'JUDGING\' \
            WHERE id = \'%d\'; \
            "
            cur.execute(updatesidStat % sid)
            logging.info('send info to other judge')
            arg = judgeLanguage + " " + pid + " " + judgeURL + " " + sid

            if os.path.exists("/var/nthuoj/outsideConnection/sendToOtherJudge.sh"):
                handle = popen("/var/nthuoj/outsideConnection/sendToOtherJudge.sh " + 
                               arg + " & >> output.html", "w")
                pclose(handle)
            else :
                logging.info('Send to other judge ERROR(No Judge)')

        sidQuery = cur.fetchone()
    time.sleep(1)

